crates/ra_syntax/src/ast/tokens.rs

```diff
@@ -172,3 +172,327 @@ impl RawString {
         Some(range + contents_range.start())
     }
 }
+
+#[derive(Debug)]
+pub enum FormatSpecifier {
+    Open,
+    Close,
+    Integer,
+    Identifier,
+    Colon,
+    Fill,
+    Align,
+    Sign,
+    NumberSign,
+    Zero,
+    DollarSign,
+    Dot,
+    Asterisk,
+    QuestionMark,
+}
+
+pub trait HasFormatSpecifier: AstToken {
+    fn lex_format_specifier<F>(&self, callback: &mut F)
+    where
+        F: FnMut(TextRange, FormatSpecifier),
+    {
+        let src = self.text().as_str();
+        let initial_len = src.len();
+        let mut chars = src.chars();
+
+        while let Some(first_char) = chars.next() {
+            match first_char {
+                '{' => {
+                    // Format specifier, see syntax at https://doc.rust-lang.org/std/fmt/index.html#syntax
+                    if chars.clone().next() == Some('{') {
+                        // Escaped format specifier, `{{`
+                        chars.next();
+                        continue;
+                    }
+
+                    let start = initial_len - chars.as_str().len() - first_char.len_utf8();
+                    let end = initial_len - chars.as_str().len();
+                    callback(
+                        TextRange::from_to(TextUnit::from_usize(start), TextUnit::from_usize(end)),
+                        FormatSpecifier::Open,
+                    );
+
+                    let next_char = if let Some(c) = chars.clone().next() {
+                        c
+                    } else {
+                        break;
+                    };
+
+                    // check for integer/identifier
+                    match next_char {
+                        '0'..='9' => {
+                            // integer
+                            read_integer(&mut chars, initial_len, callback);
+                        }
+                        'a'..='z' | 'A'..='Z' | '_' => {
```

Comment:
Do format specifiers allow unicode identifiers on nightly?

Reply:
I believe they are actually allowed even on stable (which can be considered a bug):

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d9ec6e983b041ee8937c6b6900734480

Reply:
(for those who can't read Russian, it means "nonsense" according to google translate)

Reply:
Hm, I would rather translate that as "wow". [A picture](https://gif.cmtt.space/3/paper-media/fd/92/e5/0479e08f8e852d.png) is worth a thousand words.

Reply:
I based the code on https://doc.rust-lang.org/reference/identifiers.html which says identifiers are ASCII only. Escape sequences work on stable as well: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7ad2018ba9899b4364a042eeb7dd4789
That means one has to interleave escape sequences and format specifier lexing.

Reply:
I have updated the PR to support escaped sequences and Unicode literals